packages/message-manager/src/AbstractMessageManager.ts

```diff
@@ -179,13 +195,26 @@ export abstract class AbstractMessageManager<
 
   /**
    * Adds a passed Message to this.messages, and calls this.saveMessageList() to save
-   * the unapproved Messages from that list to this.messages.
+   * the unapproved Messages from that list to this.messages. If a security provider
+   * is provided, the method will check if the Message is malicious or not.
    *
    * @param message - The Message to add to this.messages.
    */
   addMessage(message: M) {
-    this.messages.push(message);
-    this.saveMessageList();
+    if (this.securityProviderRequest) {
+      this.securityProviderRequest(message, message.type).then(
+        (securityProviderResponse) => {
+          this.messages.push({
```

Comment:
No need to duplicate this, we could simply always use the returned message from the new `checkSecurity` method and push and save once.

Reply:
I see now this was likely to delay the side effects of pushing and saving until the async work was done.

As mobile and extension appear to only use `addUnapprovedMessageAsync` which is already returning a promise, one option is to make `addUnapprovedMessage` and `addMessage` also return promises so we can `await` them and simplify the flow.

The suffix `Async` then gets a bit confusing as it doesn't mean it's a synchronous method, but rather it doesn't wait for the message to be processed.